File: src/main/java/network/matic/dagger/Room.java
package network.matic.dagger;

import network.matic.dagger.exceptions.DaggerException;

public class Room {
    private RoomType roomType;
    private Dagger dagger;

    public Room(Dagger dagger, RoomType roomType) throws DaggerException {
        if (dagger == null) {
            throw new DaggerException("`dagger` object is required");
        }

        if (roomType == null) {
            throw new DaggerException("`room` is required");
        }

        this.dagger = dagger;
        this.roomType = roomType;
    }

    public RoomType getRoomType() {
        return roomType;
    }

    public Dagger getDagger() {
        return dagger;
    }

    public Room on(String eventName, Listener listener) throws DaggerException {
        this.dagger.on(this.roomType.toString() + ":" + eventName, listener);
        return this;
    }

    public Room off(String eventName, Listener listener) throws DaggerException {
        this.dagger.off(this.roomType.toString() + ":" + eventName, listener);
        return this;
    }
}

File: src/main/java/network/matic/dagger/Room.kt
package network.matic.dagger

import network.matic.dagger.exceptions.DaggerException

class Room(dagger: Dagger?, roomType: RoomType?) {
    val roomType: RoomType
    val dagger: Dagger

    @Throws(DaggerException::class)
    fun on(eventName: String, listener: Listener?): Room {
        dagger.on("$roomType:$eventName", listener)
        return this
    }

    @Throws(DaggerException::class)
    fun off(eventName: String, listener: Listener?): Room {
        dagger.off("$roomType:$eventName", listener)
        return this
    }

    init {
        if (dagger == null) {
            throw DaggerException("`dagger` object is required")
        }
        if (roomType == null) {
            throw DaggerException("`room` is required")
        }
        this.dagger = dagger
        this.roomType = roomType
    }
}
